pkg/costattribution/manager.go

```diff
@@ -0,0 +1,236 @@
+	orderedLables := slices.Clone(labels)
+	slices.Sort(orderedLables)
+
+	tracker = newActiveSeriesTracker(userID, orderedLables, maxCardinality, cooldownDuration, m.logger)
+	m.activeTrackersByUserID[userID] = tracker
+	return tracker
+}
+
+func (m *Manager) Collect(out chan<- prometheus.Metric) {
+	m.stmtx.RLock()
+	for _, tracker := range m.sampleTrackersByUserID {
+		tracker.Collect(out)
+	}
+	m.stmtx.RUnlock()
+
+	m.atmtx.RLock()
+	for _, tracker := range m.activeTrackersByUserID {
+		tracker.Collect(out)
+	}
+	m.atmtx.RUnlock()
+}
+
+func (m *Manager) Describe(chan<- *prometheus.Desc) {
+	// Describe is not implemented because the metrics include dynamic labels. The Manager functions as an unchecked exporter.
+	// For more details, refer to the documentation: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+}
+
+func (m *Manager) deleteSampleTracker(userID string) {
+	m.stmtx.Lock()
+	delete(m.sampleTrackersByUserID, userID)
+	m.stmtx.Unlock()
+}
+
+func (m *Manager) deleteActiveTracker(userID string) {
+	m.atmtx.Lock()
+	delete(m.activeTrackersByUserID, userID)
+	m.atmtx.Unlock()
+}
+
+func (m *Manager) updateTracker(userID string) (*SampleTracker, *ActiveSeriesTracker) {
+	if !m.enabledForUser(userID) {
+		m.deleteSampleTracker(userID)
+		m.deleteActiveTracker(userID)
+		return nil, nil
+	}
+
+	st := m.SampleTracker(userID)
+	at := m.ActiveSeriesTracker(userID)
+	lbls := slices.Clone(m.limits.CostAttributionLabels(userID))
+
+	// sort the labels to ensure the order is consistent
+	slices.Sort(lbls)
+
+	// if the labels have changed or the max cardinality or cooldown duration have changed, create a new tracker
+	newMaxCardinality := m.limits.MaxCostAttributionCardinalityPerUser(userID)
+	newCooldownDuration := m.limits.CostAttributionCooldown(userID)
+
+	if !st.hasSameLabels(lbls) || st.maxCardinality != newMaxCardinality || st.cooldownDuration != newCooldownDuration {
+		m.stmtx.Lock()
+		st = newSampleTracker(userID, lbls, newMaxCardinality, newCooldownDuration, m.logger)
+		m.sampleTrackersByUserID[userID] = st
+		m.stmtx.Unlock()
+	}
+
+	if !at.hasSameLabels(lbls) || at.maxCardinality != newMaxCardinality || st.cooldownDuration != newCooldownDuration {
+		m.atmtx.Lock()
+		at = newActiveSeriesTracker(userID, lbls, newMaxCardinality, newCooldownDuration, m.logger)
+		m.activeTrackersByUserID[userID] = at
+		m.atmtx.Unlock()
+	}
+
+	return st, at
+}
+
+func (m *Manager) purgeInactiveAttributionsUntil(deadline time.Time) error {
+	m.stmtx.RLock()
+	userIDs := make([]string, 0, len(m.sampleTrackersByUserID))
+	for userID := range m.sampleTrackersByUserID {
+		userIDs = append(userIDs, userID)
+	}
+	m.stmtx.RUnlock()
+
+	for _, userID := range userIDs {
+		st, at := m.updateTracker(userID)
+		if st == nil && at == nil {
+			continue
+		}
+
+		invalidKeys := st.inactiveObservations(deadline)
+		for _, key := range invalidKeys {
+			st.cleanupTrackerAttribution(key)
+		}
```

Comment:
This is very inefficient because you have to take a write mutex for each observation to cleanup.

OTOH, you're not checking again the observations before cleaning them up.

I would suggest you to delegate this logic to the tracker:

```suggestion
		st.cleanupInactiveObservations(deadline)
```

Reply:
addressed here 41e9f477ba11a30db851e89fd8524b2dab0b39b4